File: go/floodgate.go
// SPDX-License-Identifier: MIT
package geyserlite

import (
	"crypto/rand"
	"fmt"
)

// GenerateFloodgateKey returns 16 random bytes suitable as a Floodgate
// AES-128 key.
//
// The upstream Geyser README's openssl example using
//
//	openssl genpkey -algorithm RSA
//
// is wrong — that produces an RSA private key (~1700 bytes). Floodgate uses
// AES-128, which is exactly 16 random bytes. Cf. Gate's
// pkg/edition/bedrock/geyser/floodgate/cipher.go:38 which checks
// len(key) ∈ {16, 24, 32}.
func GenerateFloodgateKey() ([]byte, error) {
	key := make([]byte, 16)
	if _, err := rand.Read(key); err != nil {
		return nil, fmt.Errorf("geyserlite: generate floodgate key: %w", err)
	}
	return key, nil
}

File: rust/src/floodgate.rs
// SPDX-License-Identifier: MIT

/// Generate a Floodgate AES-128 key (16 random bytes).
///
/// The upstream Geyser README's openssl example using
/// `openssl genpkey -algorithm RSA` is wrong — that's an RSA private key.
/// Floodgate uses AES-128, which is exactly 16 random bytes. Cf. Gate's
/// `pkg/edition/bedrock/geyser/floodgate/cipher.go:38` which checks
/// `len(key) ∈ {16, 24, 32}`.
pub fn generate_floodgate_key() -> Vec<u8> {
    use rand::TryRngCore;
    let mut buf = vec![0u8; 16];
    rand::rngs::OsRng
        .try_fill_bytes(&mut buf)
        .expect("OS RNG should not fail");
    buf
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn key_is_16_bytes() {
        assert_eq!(generate_floodgate_key().len(), 16);
    }

    #[test]
    fn keys_differ() {
        let a = generate_floodgate_key();
        let b = generate_floodgate_key();
        assert_ne!(a, b);
    }
}
